src/compile-middleware.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import * as babelCore from 'babel-core';
+import {RequestHandler, Request, Response} from 'express';
+import { transformResponse } from './transform-middleware';
+
+import {UAParser} from 'ua-parser-js';
+import * as parse5 from 'parse5';
+import * as dom5 from 'dom5';
+import {parse as parseContentType} from 'content-type';
+
+const babelLatest = require('babel-preset-latest');
+
+const javaScriptMimeTypes = [
+  'application/javascript',
+  'application/ecmascript',
+  'text/javascript',
+  'text/ecmascript',
+];
+
+const htmlMimeType = 'text/html';
+
+const compileMimeTypes = [
+  htmlMimeType,
+].concat(javaScriptMimeTypes);
+
+function getContentType(response: Response) {
+  const contentTypeHeader = response.getHeader('Content-Type');
+  return contentTypeHeader && parseContentType(contentTypeHeader).type;
+}
+
+function isSuccessful(response: Response) {
+  const statusCode = response.statusCode;
+  return (statusCode >= 200 && statusCode < 300);
+}
+
+export const babelCompile: RequestHandler = transformResponse({
+  shouldTransform(request: Request, response: Response) {
+    return isSuccessful(response) &&
+        compileMimeTypes.indexOf(getContentType(response)) >= 0;
+  },
+
+  transform(request: Request, response: Response, body: string): string {
+    const contentType = getContentType(response);
+    const uaParser = new UAParser(request.headers['user-agent']);
+    const compile = needCompilation(uaParser);
+
+    if (compile) {
+      // TODO(justinfagnani): cache compilation results
+      if (contentType === htmlMimeType) {
+        const document = parse5.parse(body);
+        const scriptTags = dom5.queryAll(document, isInlineJavaScript);
+        for (const scriptTag of scriptTags) {
+          try {
+            const script = dom5.getTextContent(scriptTag);
+            const compiledScriptResult = babelCore.transform(script, {
+              presets: [babelLatest],
+            });
+            dom5.setTextContent(scriptTag, compiledScriptResult.code);
```

Comment:
I'd recommend pulling the babelCore call into a `compile` function, just because it's highly likely that we'll want to tweak it over time, and we definitely don't want inline vs separate file compilation to drift